### ISSUE TITLE
Should subtle provide generic impls of Conditionally{Assignable, Negatable, Swappable}?

### ISSUE THREAD

**Issue by Pratyush** [dalek-cryptography/subtle#25]
Should `subtle` provide generic-impls  at all? I'm trying to use `subtle` in a library of my own at the moment, and it's creating issues. For example, if I have the following type:
```rust
#[derive(Copy)]
pub struct Thing<T>(pub T)
```
Then I can impl `ConditionallySelectable` for it, but to impl `ConditionallySwappable`, I have to have my own feature `generic-impls2` (different name for disambiguation) and then do:
```rust
#[cfg(not(feature = "generic-impls2"))]
impl<T: ConditionallySwappable> ConditionallySwappable for Thing<T> {
    fn conditional_swap(...) {
        self.0.conditional_swap(&mut other.0, choice);
    }
}
```
Now, I could impl `ConditionallySwappable` for `Thing` with an additional `Copy` bound on `T`, but that's unnecessarily restrictive and also, IMO, not correct: swapping `Thing` should only require the inner type to also be `ConditionallySwappable`.

Using the above type is thus problematic because invoking `Thing::conditional_swap` would involve having conditional trait bounds that depend on `generic-impls2`:
```rust
#[cfg(feature = "generic-impls2")]
fn do_stuff<T: ConditionallySwappable + Copy>(val: T) { 
    let something = Thing(T);
    something.conditional_swap(...);
}

#[cfg(not(feature = "generic-impls2"))]
fn do_stuff<T: ConditionalSwap>(val: T) { 
    let something = Thing(T);
    something.conditional_swap(...);
}
```

PS: Apologies for the burst of issues I've filed over the past couple of days!

**Comment by hdevalence** [comment]
No need to apologize for filing issues, they're good issues!

Your comment https://github.com/dalek-cryptography/subtle/issues/24#issue-344710470 seems relevant to this one too, linking to it so it's cross-referenced.

**Comment by Pratyush** [comment]
1) That `Copy` bound is unnecessary only for the generic impl of `ConditionallyAssignable`; generically impl'ing `ConditionallySwappable` seems to require making a temporary copy of one of the inputs, and so requires `Copy` (or perhaps some trickery with unsafe stuff).

2) I need to provide a conditional impl because whenever `Thing` is `Copy`, my unconditional impl would conflict with the one provided in subtle under `generic-impls`. Hence `rustc` disallows the blanket impl for `Thing`.

**Comment by ebfull** [comment]
`generic-impls` (as implemented) violates an invariant that Rust features are supposed to preserve: enabling a feature is reverse compatible with disabling it, so the API must not change. This contradicts `impl<T> Trait for T` impls for e.g. `ConditionallyAssignable`; when `generic-impls` is enabled the orphan rules will suddenly make a separate crate in your dependency tree fail to compile if it has an implementation of the trait.

As far as I can tell, you can provide a simple API like this:

```rust
trait ConditionallySelectable {
    fn conditional_select(a: &Self, b: &Self, choice: Choice) -> Self;
    
    // Provided impls:
    fn conditional_swap(&mut self, other: &mut Self, choice: Choice)
        where Self: Copy
    {
        // ...
    }
    
    fn conditional_assign(&mut self, other: &Self, choice: Choice)
    {
        // ...
    }
    
    fn conditional_negate(&mut self, choice: Choice)
    {
        // ...
    }
}
```

If people want to overload the provided impls for performance optimizations they can do so, or if they want to do something exotic they can make a subtrait or something.

**Comment by isislovecruft** [comment]
This should be fixed by #29.
